File: api/schemas/admin_clients.py
"""Request/response schemas for the superuser admin Clients endpoints."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator

from api.schemas.kyc import KycStatusResponse


class AdminClientItem(BaseModel):
    organization_id: int
    organization_name: str
    owner_user_id: Optional[int] = None
    owner_email: Optional[str] = None
    owner_provider_id: Optional[str] = None
    created_at: Optional[datetime] = None
    voicelink_status: Optional[str] = None
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None
    has_voicelink_config: bool = False
    did_number: Optional[str] = None
    # Live reconciliation against VoiceLink (GET /v1/reseller/clients):
    # "active" (exists in VoiceLink) | "missing" | "unconfigured" (reseller
    # creds unset) | "unknown" (reseller lookup failed → stored status shown).
    live_state: str = "unknown"
    live_client_id: Optional[str] = None
    # Remaining call-seconds balance; None = unmetered (unlimited).
    credits_seconds_remaining: Optional[int] = None


class AdminClientsListResponse(BaseModel):
    clients: List[AdminClientItem]


class RetryProvisionRequest(BaseModel):
    """A NEW VoiceLink password — client passwords are never stored locally."""

    password: str

    @field_validator("password")
    @classmethod
    def password_min_length(cls, v: str) -> str:
        if len(v) < 8:
            raise ValueError("Password must be at least 8 characters")
        return v


class RetryProvisionResponse(BaseModel):
    voicelink_status: str
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None


class CreateClientRequest(BaseModel):
    """Optional password override for one-click create.

    Normally omitted — the endpoint reuses the org's stored (encrypted)
    signup password. A password is only supplied for legacy orgs that have no
    stored secret.
    """

    password: Optional[str] = None

    @field_validator("password")
    @classmethod
    def password_min_length(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and len(v) < 8:
            raise ValueError("Password must be at least 8 characters")
        return v


class CreateClientResponse(BaseModel):
    action: str  # "linked" (already existed) | "created"
    voicelink_status: str
    voicelink_client_id: Optional[str] = None
    voicelink_username: Optional[str] = None
    voicelink_error: Optional[str] = None


class AssignDidRequest(BaseModel):
    did_number: str = Field(
        ...,
        min_length=1,
        description="DID in its VoiceLink-registered form (e.g. 919484959244)",
    )
    client_id: Optional[str] = Field(
        default=None,
        description=(
            "VoiceLink client id to stamp on the configuration; defaults to "
            "the org's provisioned voicelink_client_id"
        ),
    )


class AssignDidResponse(BaseModel):
    configuration_id: int
    created: bool
    did_number: str
    client_id: Optional[str] = None


class GrantCreditsRequest(BaseModel):
    """Top-up for a metered org's call-credits balance (1 credit = 1 minute)."""

    minutes: int = Field(
        ...,
        ge=1,
        le=100_000,
        description="Minutes of call credit to grant (converted to seconds).",
    )


class GrantCreditsResponse(BaseModel):
    organization_id: int
    granted_seconds: int
    # Balance after the grant; never None here (unmetered orgs are rejected).
    credits_seconds_remaining: Optional[int] = None


class AdminKycStatusResponse(KycStatusResponse):
    """Per-org KYC status for the admin Clients view.

    Same shape as the self-serve ``GET /kyc/status`` response plus a
    ``status`` discriminator:

    - ``ok`` — the org's VoiceLink ``client_id`` resolved and the KYC status
      was fetched from VoiceLink.
    - ``no_client`` — the org has no resolvable VoiceLink client id (KYC
      would act on the reseller's own account, so we don't fetch).
    - ``disabled`` — reseller credentials are not configured.
    """

    status: str = "ok"
    client_id: Optional[str] = None

File: ui/src/lib/adminClients.ts
/**
 * Small fetch wrapper for the superuser admin Clients routes
 * (`/api/v1/admin/clients*`).
 *
 * These routes are not part of the generated OpenAPI client yet, so this
 * mirrors the client's base-URL resolution and Bearer-token convention
 * (same pattern as lib/kyc.ts).
 */

export type VoiceLinkLiveState =
  | "active"
  | "missing"
  | "unconfigured"
  | "unknown";

export interface AdminClient {
  organization_id: number;
  organization_name: string;
  owner_user_id?: number | null;
  owner_email?: string | null;
  owner_provider_id?: string | null;
  created_at?: string | null;
  voicelink_status?: string | null;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
  has_voicelink_config: boolean;
  did_number?: string | null;
  // Live reconciliation against VoiceLink.
  live_state: VoiceLinkLiveState;
  live_client_id?: string | null;
  // Remaining call-seconds balance; null = unmetered (unlimited).
  credits_seconds_remaining?: number | null;
}

export interface AdminClientsListResult {
  clients: AdminClient[];
}

export interface RetryProvisionResult {
  voicelink_status: string;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
}

export interface CreateClientResult {
  action: string; // "linked" | "created"
  voicelink_status: string;
  voicelink_client_id?: string | null;
  voicelink_username?: string | null;
  voicelink_error?: string | null;
}

export interface AssignDidBody {
  did_number: string;
  client_id?: string;
}

export interface AssignDidResult {
  configuration_id: number;
  created: boolean;
  did_number: string;
  client_id?: string | null;
}

export interface GrantCreditsResult {
  organization_id: number;
  granted_seconds: number;
  credits_seconds_remaining?: number | null;
}

// "ok" = fetched from VoiceLink | "no_client" = org has no VoiceLink client
// id | "disabled" = reseller credentials unset on the backend.
export type AdminKycState = "ok" | "no_client" | "disabled";

export interface AdminClientKycStatus {
  status: AdminKycState;
  enabled: boolean;
  client_id_configured?: boolean;
  has_voicelink_config?: boolean;
  client_id?: string | null;
  kyc_status?: string | null;
  pan_verified?: boolean | null;
  aadhaar_verified?: boolean | null;
  gst_verified?: boolean | null;
  is_complete?: boolean | null;
  current_step?: number | string | null;
  account_type?: string | null;
}

function backendUrl(): string {
  return (
    process.env.NEXT_PUBLIC_BACKEND_URL ||
    (typeof window !== "undefined" ? window.location.origin : "")
  );
}

function detailFromBody(body: unknown): string {
  const e = body as { detail?: unknown };
  if (typeof e?.detail === "string") return e.detail;
  if (Array.isArray(e?.detail) && e.detail.length > 0) {
    const first = e.detail[0] as { msg?: string };
    if (first?.msg) return first.msg;
  }
  return "Request failed";
}

async function adminFetch<T>(
  token: string,
  path: string,
  init?: RequestInit,
): Promise<T> {
  const res = await fetch(`${backendUrl()}/api/v1/admin/clients${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
      ...(init?.headers ?? {}),
    },
  });
  let body: unknown = {};
  try {
    body = await res.json();
  } catch {
    // Non-JSON response body — fall through to the generic error below.
  }
  if (!res.ok) throw new Error(detailFromBody(body));
  return body as T;
}

export const listAdminClients = (token: string) =>
  adminFetch<AdminClientsListResult>(token, "");

export const retryProvisionClient = (
  token: string,
  organizationId: number,
  password: string,
) =>
  adminFetch<RetryProvisionResult>(
    token,
    `/${organizationId}/retry-provision`,
    { method: "POST", body: JSON.stringify({ password }) },
  );

export const createClientForOrg = (
  token: string,
  organizationId: number,
  password?: string,
) =>
  adminFetch<CreateClientResult>(token, `/${organizationId}/create`, {
    method: "POST",
    body: JSON.stringify(password ? { password } : {}),
  });

export const assignDidToClient = (
  token: string,
  organizationId: number,
  body: AssignDidBody,
) =>
  adminFetch<AssignDidResult>(token, `/${organizationId}/assign-did`, {
    method: "POST",
    body: JSON.stringify(body),
  });

export const grantCreditsToClient = (
  token: string,
  organizationId: number,
  minutes: number,
) =>
  adminFetch<GrantCreditsResult>(token, `/${organizationId}/grant-credits`, {
    method: "POST",
    body: JSON.stringify({ minutes }),
  });

export const getClientKycStatus = (token: string, organizationId: number) =>
  adminFetch<AdminClientKycStatus>(token, `/${organizationId}/kyc-status`);
